lib/block-supports/duotone.php

```diff
@@ -445,15 +442,44 @@ function gutenberg_render_duotone_support( $block_content, $block ) {
 		return $block_content;
 	}
 
-	$colors          = $block['attrs']['style']['color']['duotone'];
-	$filter_key      = is_array( $colors ) ? implode( '-', $colors ) : $colors;
-	$filter_preset   = array(
-		'slug'   => wp_unique_id( sanitize_key( $filter_key . '-' ) ),
-		'colors' => $colors,
-	);
-	$filter_property = gutenberg_get_duotone_filter_property( $filter_preset );
-	$filter_id       = gutenberg_get_duotone_filter_id( $filter_preset );
+	// Possible values for duotone attribute:
+	// 1. array('#000000', '#ffffff').
+	// 2. 'preset-slug'.
+	// 3. 'unset' (remove filter).
+	$duotone_attr = $block['attrs']['style']['color']['duotone'];
```

Comment:
I do think it's worthwhile in this PR to keep the styles separate from the presets rather than moving it to a follow-up.

Conceptually, I think the `style` attribute was supposed to represent inline styles that were applied to a block. We can't just set `<div class="wp-block-image" style="filter:url('data:image/svg+xml...');">...</div>` because kses strips SVGs and data URL filters don't work in Safari anyway. But conceptually, that's what it means.

Whereas, the top-level attributes (`textColor`, `gradient`, etc.) represented presets that were usually applied with a class on the block wrapper. Again, we can't just apply a class to the wrapper because `filter` is a non-inherited CSS property that we want to apply to a subset of the block.

So I think a top-level `duotone` attribute is more appropriate for saving the presets to.

Reply:
I agree it's needed...but that's going to involve:

- deprecations
- additional refactoring to work with the new attribute whilst maintaining the backwards compatibility

...I still think I would be better to make an immediate follow up and avoid bloating this PR.

As soon as I feel I would need to use the word `and` in a PR description to accurately reflect its purpose it's become too large.

Is there a reason it _has_ to be done in _this_ PR?

Reply:
Why do you think it would need deprecations or additional backwards compatibility code?

The way I see it, old blocks with the custom colors saved continue to work like they did before. Nothing here gets touched.

New blocks that select a preset have a new attribute that unsets the `style.color.duotone` attribute and sets the new `duotone` attribute.

I think that it should be this PR so that we don't have to deal with backwards compatibility when switching over to storing custom filters in `style` and presets in `duotone`.

Reply:
Eeeek. Too late. But I'll do a follow up 😶

Reply:
So we should end up with

- `attributes.filters.duotone` - stores duotone presets only.
- `attributes.style.color.duotone` - stores custom duotone only.

Is that right?